src/backend/storage/ipc/procarray.c

```diff
@@ -4929,38 +4929,188 @@ GetSessionIdByPid(int pid)
  * Set the destination group slot or group id in PGPROC, and send a signal to the proc.
  * slot is NULL on QE.
  */
-void
-ResGroupSignalMoveQuery(int sessionId, void *slot, Oid groupId)
+bool
+ResGroupMoveSignalTarget(int sessionId, void *slot, Oid groupId, bool executor)
 {
 	pid_t pid;
 	BackendId backendId;
 	ProcArrayStruct *arrayP = procArray;
+	bool sent = false;
+	bool found = false;
+
+	Assert(Gp_role == GP_ROLE_DISPATCH || Gp_role == GP_ROLE_EXECUTE);
+	Assert(groupId != InvalidOid);
+
+	if (Gp_role == GP_ROLE_EXECUTE)
+		executor = true;
 
 	LWLockAcquire(ProcArrayLock, LW_SHARED);
 	for (int i = 0; i < arrayP->numProcs; i++)
 	{
 		volatile PGPROC *proc = &allProcs[arrayP->pgprocnos[i]];
 		if (proc->mppSessionId != sessionId)
 			continue;
+		/* 
+		 * Before, we didn't distinguish entrydb processes from target.
+		 * There is a case with enrtydb executors when we can send a signal to
+		 * target process only, but not to entrydb executor process or
+		 * vice versa.
+		 * As a mediocre solution we assume executors (including entrydb) not
+		 * using queryCommandId.
+		 */
+		if ((proc->queryCommandId <= -1) ^ executor)
```

Comment:
Although I don't have any alternatives to suggest, don’t like the idea of relying on the queryCommandId identifier. There is not a zero percent chance someday we will need it on executors as well

Reply:
The guard is [here](https://github.com/arenadata/gpdb/pull/300/files/665710722804bda1ad7248a8214c70f6ee29dd06#diff-14f7111e45936f7f6b0fbccf732d91ab7957e461b8a3e641381e1b003f113a18R4711-R4716).

Reply:
Agree, although still don't like it :)

Reply:
The word "mediocre" points to this too.

Reply:
Hm... And what do you think about mppIsWriter property?
It is seems, it is [always true](https://github.com/arenadata/gpdb/blob/ADBDEV-2419-2/src/backend/storage/lmgr/proc.c#L490-L493) for dispatcher role
And is always false for [entrydb](https://github.com/arenadata/gpdb/blob/ADBDEV-2419-2/src/backend/cdb/cdbutil.c#L805) (for [utility role as well](https://github.com/arenadata/gpdb/blob/ADBDEV-2419-2/src/backend/cdb/cdbvars.c#L471-L472))

Reply:
OK. So it should be less tricky, right?

Reply:
I suppose, it is will look almost the same but slightly less mediocre (there are some places in code where the same approach is used), what is your opinion?

Reply:
No so handy as `queryCommandId`, but OK.

Reply:
Switched to use of `mppIsWriter`. I proved the behavior by debugging.
Assert() in handler removed.
Thanks.